### PR TITLE
cmdline app: add provider commandline options.

### PR DESCRIPTION
**Pull request by paulidale**
Compare with #11160.

Add a -provider option to allow providers to be loaded. This option can be
specified multiple times.

Add a -provider_path option to allow the path to providers to be specified.

- [ ] documentation is added or updated
- [ ] tests are added or updated

### CODE REVIEW

**Comment by paulidale** [comment]
WIP because:

1. only two of nearly all apps have been updated.
1. no tests

**Comment by t8m** [comment]
I like this approach more.

**Comment by paulidale** [comment]
Travis isn't relevant, Windows is failing on `setenv` and `unsetenv` which need a better solution.

**Comment by paulidale** [comment]
Moving this out of WIP.
Assuming the CI loops succeed.

I've deliberately not delved into the intricacies of the interactive command line application.

1. Should each entered command get its own library context or should one be shared?  If shared, unload needs to be supported.
1. Should provider options be accepted on the initial openssl command line (without a command name)?

**Comment by richsalz** [comment]
> I've deliberately not delved into the intricacies of the interactive command line application.

Many things are not reset properly in "interactive mode" as I recall.  It would be nice to deprecate this.

**Comment by paulidale** [comment]
ping.

**Comment by paulidale** [comment]
Ping for review.

File: apps/crl.c [hunk 1]
```diff
@@ -193,6 +194,10 @@ int crl_main(int argc, char **argv)
         case OPT_MD:
             if (!opt_md(opt_unknown(), &digest))
                 goto opthelp;
+        case OPT_PROV_CASES:
```
Comment (t8m) [inline apps/crl.c]: missing break; above.

File: apps/genpkey.c [hunk 2]
```diff
@@ -131,6 +134,10 @@ int genpkey_main(int argc, char **argv)
                 BIO_printf(bio_err, "%s: cipher mode not supported\n", prog);
                 goto end;
             }
+        case OPT_PROV_CASES:
```
Comment (t8m) [inline apps/genpkey.c]: missing break

File: crypto/provider_core.c [hunk 3]
```diff
@@ -384,6 +387,27 @@ int ossl_provider_add_parameter(OSSL_PROVIDER *prov,
...
 
+int OSSL_PROVIDER_set_default_search_path(OPENSSL_CTX *libctx, const char *path)
+{
+    struct provider_store_st *store;
+    char *p = NULL;
+
+    if (path != NULL) {
+        p = OPENSSL_strdup(path);
+        if (p == NULL)
+            return 0;
```
Comment (t8m) [inline crypto/provider_core.c]: Should there be an error added?

File: doc/man1/openssl.pod [hunk 4]
```diff
@@ -755,6 +755,26 @@ This file can be used in a subsequent command invocation.
 
 =back
 
+=head2 Provider Options
+
+All cryptographic operations in OpenSSL 3.0 are provider based.  These options
```
Comment (t8m) [inline doc/man1/openssl.pod]: This is misleading. Even if we ignore legacy codepaths in EVP that will be removed before release, there are the engines and the legacy APIs.

File: doc/man3/OSSL_PROVIDER.pod [hunk 5]
```diff
@@ -39,6 +43,11 @@ L<OPENSSL_CTX(3)> for further details.
 
 =head2 Functions
 
+OSSL_PROVIDER_set_default_search_path() specifies the default search B<path>
+that is to be used for looking for providers in the specified B<libctx>.
+If left unspecified, an environment variable and a fall back default value will
```
Comment (t8m) [inline doc/man3/OSSL_PROVIDER.pod]: fall-back?

**Comment by paulidale** [comment]
Rebased to address merge conflict (renaming of .conf to .cnf in documentation).

**Comment by paulidale** [comment]
Merged to master.  Thanks.

Travis failure was the usual timeout.

**Pull request closed without merging**